src/s3ql/http.py

```diff
@@ -0,0 +1,1558 @@
+        '''
+
+        log.debug('start (len=%d)', len(buf))
+
+        if not self._out_remaining:
+            raise StateError('No active request with pending body data')
+
+        (method, path, remaining) = self._out_remaining
+        if remaining is WAITING_FOR_100c:
+            raise StateError("can't write when waiting for 100-continue")
+
+        if len(buf) > remaining:
+            raise ExcessBodyData(
+                'trying to write %d bytes, but only %d bytes pending' % (len(buf), remaining)
+            )
+
+        try:
+            await self._co_send(buf)
+        except ConnectionClosed:
+            # If the server closed the connection, we pretend that all data
+            # has been sent, so that we can still read a (buffered) error
+            # response.
+            self._out_remaining = None
+            self._pending_requests.append((method, path, None))
+            raise
+
+        len_ = len(buf)
+        if len_ == remaining:
+            log.debug('body sent fully')
+            self._out_remaining = None
+            self._pending_requests.append((method, path, None))
+        else:
+            self._out_remaining = (method, path, remaining - len_)
+
+        log.debug('done')
+
+    def response_pending(self):
+        '''Return `True` if there are still outstanding responses
+
+        This includes responses that have been partially read.
+        '''
+
+        return self._sock is not None and len(self._pending_requests) > 0
+
+    def read_response(self):
+        '''placeholder, will be replaced dynamically'''
+        return eval_coroutine(self.co_read_response(), self.timeout)
+
+    async def co_read_response(self):
+        '''Read response status line and headers
+
+        Return a `HTTPResponse` instance containing information about response
+        status, reason, and headers. The response body data must be retrieved
+        separately (e.g. using `.read` or `.readall`).
+        '''
+
+        log.debug('start')
+
+        if len(self._pending_requests) == 0:
+            raise StateError('No pending requests')
+
+        if self._in_remaining is not None:
+            raise StateError('Previous response not read completely')
+
+        (method, path, body_size) = self._pending_requests[0]
+
+        # Need to loop to handle any 1xx responses
+        while True:
+            (status, reason) = await self._co_read_status()
+            log.debug('got %03d %s', status, reason)
+
+            hstring = await self._co_read_header()
+            header = email.message_from_string(hstring, policy=email.policy.HTTP)
+
+            if status < 100 or status > 199:
+                break
+
+            # We are waiting for 100-continue
+            if body_size is not None and status == 100:
+                break
+
+        # Handle (expected) 100-continue
+        if status == 100:
+            assert self._out_remaining == (method, path, WAITING_FOR_100c)
```

Comment:
`assert` isn't called when running Python in optimized mode. Is that desired here?